assets/js/plugins.coffee

```diff
@@ -0,0 +1,99 @@
+if module?
+  global._ = require('lodash')
+  global.Logger = require('./logger.coffee')
+  global.DependencyGraph = require('dependencies-online')
+
+(() ->
+  class PluginDatabaseAccess
+    constructor: (@data, @pluginName) ->
+
+    getVersion: () ->
+      @data.store.getPluginVersion @pluginName
+    setVersion: (version) ->
+      @data.store.setPluginVersion @pluginName, version
+    setRowData: (row, args...) ->
+      if args.length == 1
+        [key, [data]] = ['data', args]
+      else if args.length == 2
+        [key, data] = args
+      @data.store.setPluginRowData @pluginName, row.id, key, data
+    getRowData: (row, key='data') ->
+      @data.store.getPluginRowData @pluginName, row.id, key
+    setGlobalData: (key, args...) ->
+      if args.length == 1
+        [key, [data]] = ['data', args]
+      else if args.length == 2
+        [key, data] = args
+      @data.store.setPluginGlobalData @pluginName, key, data
+    getGlobalData: (key='data') ->
+      @data.store.getPluginGlobalData @pluginName, key
+
+    transformRowData: (row, args...) ->
+      if args.length == 1
+        [key, [transform]] = ['data', args]
+      else if args.length == 2
+        [key, transform] = args
+      @setRowData row, key, (transform (@getRowData row, key))
+    transformGlobalData: (args...) ->
+      if args.length == 1
+        [key, [transform]] = ['data', args]
+      else if args.length == 2
+        [key, transform] = args
+      @setGlobalData key, (transform (@getGlobalData key))
+  
+  class PluginAPI
+    constructor: (@view) ->
+      @plugins = {}
+      @pluginClass = {}
+      @metadata = {}
+      @pluginDependencies = new DependencyGraph()
+      # Convenience accessors
+      @cursor = @view.cursor
+
+    registerPlugin: (plugin) ->
+      # Validate metadata before accepting registration
+      metadata = plugin.metadata
+      unless metadata?
+        throw "Plugin is missing all metadata"
+      required_props = ['name', 'version', 'stores_data']
+      if (_.result metadata, 'stores_data')
+        required_props.push 'data_version'
+      for prop in ['name', 'version', 'stores_data']
+        unless _.has metadata, prop
+          throw "Plugin #{plugin?.name ? "<Unnamed Plugin"} has missing metadata field #{prop}"
+
+      Logger.logger.info "Plugin #{plugin.name} registered"
+      # Store the parts of the plugin in case we want to register them later
+      @metadata[plugin.name] = plugin.metadata
+      @pluginClass[plugin.name] = plugin
+      requirements = _.result plugin.metadata, 'requirements', []
+      # Load the plugin after all its dependencies have loaded
+      @pluginDependencies.add plugin.name, requirements, () =>
+        Logger.logger.info "Plugin #{plugin.name} loading"
+        @_currentlyRegistering = plugin.name
+        @plugins[plugin.name] = new @pluginClass[plugin.name] @
+        delete @_currentlyRegistering
+        @plugins[plugin.name]._metadata = @metadata[plugin.name]
```

Comment:
Cyclic: An error is raised
Dependencies that aren't resolved: Nothing, the plugin just never loads. Could probably change this if you wanted.

The idea is that the plugins load asynchronously, yes. I think the idea is less that they rely on other plugins, and more that you say, might want them to load as soon as possible vs after the editor is visible, once we start factoring stuff into plugins a bit.